### ISSUE TITLE
Auto-submit bot should set the commit message of the squashed commit

### ISSUE THREAD

**Issue by chinmaygarde** [flutter/flutter#49388]
This came up in the Flutter Engine weekly about the behavior of the commit queue opted into via the "waiting for the tree to go green" label. @dnfield clarified that only the headline of the primary commit message is kept while the descriptions are discarded. Some commit authors put a [significant](https://github.com/flutter/engine/commit/ad582b5089256e56727b2d7306fb097078a3dcdc) [amount](https://github.com/flutter/engine/commit/c7d0fb787922cb60531c415ed7bf652afbaea552) of [effort](https://github.com/flutter/engine/commit/b454251f03f27ff422b2c197761493a9e3133909) in adding a descriptive commit message and keeping the commit message updated across amendments during the review. These messages are also useful when locally reviewing patches and blames. While linking to the PR certainly works, it is one more step to follow from the commit and it would be preferable to keep things in the Git repo instead of needing GitHub for basic code navigation tasks.

As it stands, this behavior is a hinderance to the adoption of the commit queue.

**Comment by chinmaygarde** [comment]
cc @godofredoc @dnfield

**Comment by dnfield** [comment]
See also #49091 

I think we could do this, but we'd want some kind of lint about commit messages in that case.

**Comment by chinmaygarde** [comment]
> we'd want some kind of lint about commit messages in that case

I am not opposed to this. By why is this related to the the commit queue? We have needed lints even before the commit queue when folks used the default GitHub Squash-And-Merge workflow that appended the titles of followup commits (usually meaningless) to the commit message.

**Comment by dnfield** [comment]
Sorry, I mean if we want the bot to _not_ discard messages, we should set up some kind of lint about commit messages to try to make sure we avoid garbage in the log.

**Comment by dnfield** [comment]
I also have a slightly different take on the value of the git log - while the explanations entered in those commit messgaes is really great, it seems like they might be better off in a doc or collection of docs that explain the design/features of the repo.

I'm less likely to discover them burried in the log with 100s of commits from autorollers I don't care about than if they're published somewhere else for people who want to read these things.  But I think that's a preference, not a necessity.  

At any rate, we should figure out if we can somehow lint the commit messages in a meaningful way, and if there's some standard we can agree on for them.  /cc @hixie @tvolkert @cbracken

**Comment by chinmaygarde** [comment]
> I also have a slightly different take on the value of the git log

Again. The function of the commit queue and what we as a team decide what the commit message should look like are orthogonal concerns. But we seem to have implemented a commit queue and settled on commit message convention as a side effect of implementing the commit queue instead of after discussion as a team.

**Comment by chinmaygarde** [comment]
> it seems like they might be better off in a doc or collection of docs that explain the design/features of the repo. I'm less likely to discover them burried in the log with 100s of commits from autorollers I don't care about than if they're published somewhere else for people who want to read these things. But I think that's a preference, not a necessity.

On the policy itself: Commit message descriptions elucidate the thought process of the committer and describe the issues faced/considered by the commit author **at the time the commit was made**.  This context, frozen in amber, is precisely what makes them so valuable. Design docs or other living docs serve a vital but somewhat orthogonal purpose and I am in no way against writing those as well. Besides, the implementation of the policy doesn't automatically mean folks are now forced/encouraged to write the types of docs you would like to see. If your goal is to frustrate authors into writing these docs, I vote for a kinder approach to achieving this goal.

**Comment by dnfield** [comment]
To be clear, I implemented the merge functionality that way because I was assuming that most commit messages were garbage and the good ones would still be available on GitHub - my assumption (perhaps wrong) was that we'd be more concerned with clutter in the log than needing to go to GitHub for the good stuff.

It's a fairly trivial change to land the commit message as is. It would be great though if we had some way to lint that commit messages weren't just trash before landing things though.

I'll ping this issue to the hackers chat to see if anyone has more feedback.  If there are other ideas to get feedback from the team I'm game too.

**Comment by Hixie** [comment]
I sure hope the commit messages aren't garbage!

I would expect the first comment in a PR to be the commit message. (Maybe stripping the templated part.)

**Comment by dnfield** [comment]
Here's an example of what the bot is replacing today, from a PR I currently have open:

```
* Elide tree walks

* oops

* Elide everything after mount

* ...

* another attempt

* ..

* make it work for web, clean up logic

* test

* review/tests

* remove expectations

* test interference

* review

* todo
```

The headline for that is `Elide tree walks`.

**Comment by Hixie** [comment]
I mean, you should write better commit messages... :-)

**Comment by dnfield** [comment]
haha sure, but at this point I'm used to just squash merging and not caring about anything after the first.  I also just tend to focus more on the GitHub PR page than what ends up in the git log.

**Comment by Hixie** [comment]
Yeah, me too. Hence my original suggestion of using the first comment in the PR.

**Comment by cbracken** [comment]
I'm not sure what the best strategy is here, but I know I put probably way more time than I should into writing good commit messages after a couple incidents where I'd gone back looking for rationale for something during an archaeological dig and found my own previous commit messages lacking a crucial detail. Reverts without rationale in particular drive me absolutely bananas. I [try](https://github.com/flutter/engine/commit/23df3e6bcbcbbd4b475249f94978e66c6be127f5) to ensure there's context for whoever attempts the re-land.

In general, I expect using the initial commit is sufficient. In most cases, I see stuff like "format" on further commits.

**Comment by Piinks** [comment]
The Gold team relies on the associated PR being at the end of the headline in order to handle tryjob ingestion on master. e.g. `Changed licenses (#XXXXX)`
As for the description, I once forgot to clean up the various commit messages from a long-term project, and they ended up being individually weaved throughout the commit history rather than when the commit was actually made on master. In that case I think it made very little sense from an archaeological standpoint to try to identify what change a specific commit message was associated with. I agree with @dnfield  that a linter would be beneficial for a clearer history.

**Comment by chinmaygarde** [comment]
How about making the commit queue not consider un-squashed commits? Then, you can take the commit message as-is. Auto-squash is not a feature I particularly care about and is just adding ambiguity.

As I said, I really want to use the commit queue but also put a lot of effort into commit messages. In the past, I have reached out to folks to remind them to tweak the commit message before committing (always after the fact because the squash+merge happens after the review). But, this didn't seem to change behavior and only made me feel like a nitpicky asshole. I suppose that is not entirely surprising since there is clearly a spectrum of opinion on the value of commit messages. My point is, let's solve the "I need a commit queue" and "there is garbage in commit messages" problems separately.

**Comment by cbracken** [comment]
@Piinks wrote:
> The Gold team relies on the associated PR being at the end of the headline

I can't remember how github generates the "title" (i.e. first line) of a commit comment, but IIRC it just uses the first line of the first commit, then tacks on the PR info in brackets at the end, doesn't it? How easy would it be to keep the first line + following blank line, but replace anything after that with everything after the first line in the original commit, I wonder?

@chinmaygarde wrote:
> How about making the commit queue not consider un-squashed commits?

I'd be entirely on board with this. The only case this should affect is PRs from contributors who aren't on the team -- we'd need to remind them to rebase + squash with a decent commit message. I've run into a few people who aren't familiar enough with git to do that, but in general I've just done it for them. Extra 5 mins of work on my end, but not really a big additional cost all things considered. Either that or we write up a short guide we can point people to if they aren't familiar with `git rebase -i`.

**Comment by dnfield** [comment]
Sometimes you want to have multiple commits in your PR to make reviewing certain parts of the changes easier.

And to find out your commit wasn't merged because you forgot to `rebase -i`, and then you rebase to squash down, and then you find out the tree is red again, and you apply the label, and find out CI on your PR went red because of a flake ...  and we lose a lot of the value.

**Comment by dnfield** [comment]
I would be fine with having the bot keep the first commit message it sees (e.g. the first `*` if there's a markdown list, or the whole thing otherwise).  It sounds like that would avoid flooding the log with garbage and still keep the good stuff to me.

**Comment by cbracken** [comment]
> Sometimes you want to have multiple commits in your PR to make reviewing certain parts of the changes easier.

Requiring a manual squash + push once you've got LGTM shouldn't prevent that. e.g. my workflow is to push a new commits with each review round (and sometimes up-front if I want to break out logically-related changes for the reviewer), but do a manual squash-re-push at the end. Admittedly, that's mostly because I find GitHub's textfield sub-optimal (doesn't use a non-proportional font or auto-wrap at the 72-column mark, but it's not a huge barrier here.

> I would be fine with having the bot keep the first commit message it sees (e.g. the first * if there's a markdown list, or the whole thing otherwise).

Seems like a reasonable heuristic, though I tend to use markdown lists in my commit comments when things get long. Those typically appear mid-commit-message though, whereas squash commits should start with them on the first non-blank line after the header... I think?

**Comment by jamesderlin** [comment]
I also expected using the description in the first comment (not the first commit).  I recently made a PR that I broke into several commits for easier review, and the commit description for each individual commit described what it was doing.  Using the first commit message in that case would not have been appropriate.

There are perhaps some compromises that could be made.  Tossing out some ideas:
* If nothing else, include a link back to the GitHub pull request to make the manual process of selecting, copying, and pasting the PR number a bit less painful. (Not sure if this would spam the PR with an unnecessary "... added a commit ... that referenced this issue" message.)
* Look for a comment on the PR that explicitly states what the desired commit message should be.
* Email the committer when the tree is green instead of merging automatically if the committer chooses or if there's some ambiguity about what to do.

**Comment by dnfield** [comment]
Perhaps we could adopt some convention in the pull request body message that gets stuffed into the merge by the bot

**Comment by jmagman** [comment]
Here's a bad one:
https://github.com/flutter/flutter/pull/88391 has the title "Fix BoxDecoration crash with BorderRadiusDirectional".  fluttergithubbot merged it as `init (#88391)`.  This makes git archeology difficult.
<img width="378" alt="Screen Shot 2021-08-20 at 11 49 33 AM" src="https://user-images.githubusercontent.com/682784/130280131-e43f4e79-a037-4688-a0f7-3612240b5aef.png">

**Comment by nt4f04uNd** [comment]
I'm sorry for that being a thing, I'm used to that PR title goes into the commit name. I guess I will make the first commit to have a meaningful message from now

**Comment by dnfield** [comment]
Not your fault :) I was under the impression that the PR title ended up being the commit message too - I'm not sure if this functionality has changed though.

**Comment by jmagman** [comment]
> I'm sorry for that being a thing, I'm used to that PR title goes into the commit name. I guess I will make the first commit to have a meaningful message from now

@nt4f04uNd Not your fault at all!  I'm annoyed on your behalf, I've done it quite a few times myself.  🙂

**Comment by jmagman** [comment]
@dnfield I guess this isn't the right issue then, this is tracking the missing squashed commit messages, not that the PR title isn't in the commit message.  I'll file a new one.

**Comment by cbracken** [comment]
Ran into this again this week when I hit this again doing commit archaeology to figure out the rationale for the presence of a few lines of code. Created a new issue since I forgot this one not only existed, but I'd commented on it a bunch. Added more rationale and some proposals for potential fixes in https://github.com/flutter/flutter/issues/103540.

**Comment by pdblasi-google** [comment]
For me, I like having a descriptive history available. Especially when I see something that _looks_ superfluous or otherwise unneeded, but I'm not _certain_ because I don't have the context. My own experience and preferences have led to me usually using the following format for my first commit on a new branch:

```
<Issue/Bug/Ticket #>: Title/description of ticket

* Individual change made
  * Rationale/further details about change
* Another change
* etc.
```

Which has made creating PRs, finding related tickets, and having local access to context regarding the implementation immediately available. It seems like this (or a similar) format is expected by most online repo services (i.e. GitHub, GitLab) as they tend to use the first line of the first commit as the title of a PR by default, and the rest of that commit as a description or initial comment on the PR.

Given the various suggestions and context in this and the #103540 ticket, I'd throw my vote for making the auto-submit bot create the commit message out of the title and initial comment/description (minus boilerplate). From my perspective, that option:

1. Incentivizes good (imo) commit message structure indirectly by having PRs auto-populated with reasonable values
2. Allows for variance in commit preferences by letting the end result be changed outside of the commit history
    *  i.e. Splitting PRs into more reviewable smaller commits, but without any "overview" commit message in the history
4. Allows for variance in git skill/knowledge by not requiring single commit PRs (which lets us be more welcoming!)
5. Creates more context available locally in the git history after merging

**Comment by godofredoc** [comment]
@itsjustkevin @muditatandon this is one of the bugs about commit descriptions we were discussing this morning.

**Comment by dnfield** [comment]
Maybe the answer here is to come up with a style guide for commit descriptions and having some way for the bot to enforce that.

I still think using some part of the original PR description would be ideal, or perhaps just verifying the initial commit description. I'm just not sure what we'd do for violations if we use any of the actual commit history from the user, since we'd have to give them reasonable instructions for force pushing, which if they mess up can make things difficult to fix.

**Comment by pdblasi-google** [comment]
I think a style guide for commit descriptions (at least an enforced one) isn't a good idea. As others have brought up, there are good reasons to have commits that aren't fully descriptive of an entire branch in the _branch's_ history, but it _is_ a good thing to have after merging.

That's why I agree with your second paragraph more, in using the PR's title and description. Those things are easier to grasp, change, and to see when doing a review, so they're less likely to require hard rules that make things less welcoming for people with less experience with git's more niche features.

**Comment by cbracken** [comment]
Agreed on using the PR description (minus the boilerplate, ideally). While it does mean that anyone who's primarily command-line driven needs to go back and copy-paste any changes they make along the way into the PR description, that feels a lot easier and safer than asking contributors to rebase/amend + force push.

**Comment by stuartmorgan** [comment]
Strong +1 on PR description being preferred. For non-team contributions, we can trivially adjust the PR description to improve it (I do this regularly in flutter/plugins and flutter/packages), but we can't change the commit messages.

**Comment by gnprice** [comment]
This was done a while ago (IIRC sometime in 2023). Auto-submitted PRs since then have gotten their commit message based on the PR description, which has been great.
